Get blank forms from Google Drive using API v3

shobhitagarwal1612 opened the pull request:
In continuation of #713
Also fixes #650

shobhitagarwal1612 committed:
Adds dependency for drive v3

shobhitagarwal1612 committed:
Renames mService to mSheetsService

shobhitagarwal1612 committed:
media upload to google drive implemented

shobhitagarwal1612 committed:
media upload to google drive implemented

shobhitagarwal1612 committed:
Photo link uploading to drive

shobhitagarwal1612 committed:
Removes Picasa directory

shobhitagarwal1612 committed:
Rename gradle dependencies according to long naming convention

shobhitagarwal1612 committed:
Implements OnClickListener interface to GoogleDriveActivity

shobhitagarwal1612 committed:
Modifies directory structure of google drive uploads folder

shobhitagarwal1612 committed:
Get blank forms from drive updated according to api v3

shobhitagarwal1612 committed:
Updates license year to 2017

grzesiek2010 commented:
@shobhitagarwal1612 
Before we had to click on "My Drive" button to see files now I can see all files automatically is it an 
expected behaviour?

shobhitagarwal1612 commented:
@grzesiek2010 Earlier, we used to click the button to get the drive files even after going into the `Get blank forms`. So I called the api as soon as the user enters the activity, saving one click of the user whenever trying to reach this section of the app. 
I'll revert it back if this doesn't seems ok.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -56,104 +59,83 @@
+import pub.devrel.easypermissions.AfterPermissionGranted;
+import pub.devrel.easypermissions.EasyPermissions;
+
+//import com.google.api.services.drive.model.ParentList;
```
This can simply be removed.

shobhitagarwal1612 replied:
Done

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -315,14 +244,145 @@ public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+     * account was selected and the device currently has online access. If any
+     * of the preconditions are not satisfied, the app will prompt the user as
+     * appropriate.
+     */
```
Would be good to have a quick link to https://developers.google.com/drive/v3/web/quickstart/android like you put in other places.

shobhitagarwal1612 replied:
Is there any other place too where the reference link should be added?

lognaturel replied:
Hmm, I thought it would be helpful with some of the OAuth stuff but I wasn't finding the best reference to it. It was just coming up on a generic "set up an Android project" page. If you have a page with the code samples and some explanation, that would be a good link to have. I think you may already have linked to something like that in the sheets feature.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -526,82 +618,238 @@ public Object onRetainNonConfigurationInstance() {
+        super.onStop();
+    }
+
+    private boolean testNetwork() {
```
`isDeviceOnline` appears to be a slightly cleaner implementation of the same thing so this method and corresponding calls should be replaced.

shobhitagarwal1612 replied:
Replaced

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -526,82 +618,238 @@ public Object onRetainNonConfigurationInstance() {
+                startActivityForResult(e.getIntent(), COMPLETE_AUTHORIZATION_REQUEST_CODE);
+                return null;
+            } catch (IOException e) {
+                e.printStackTrace();
```
We are trying to move away from using `e.printStackTrace()` since it always gets lost. Can you please replace it with `Log.e` throughout the code (and it will be in turned replaced by `Timber.e` soon).

shobhitagarwal1612 replied:
Sure

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -315,14 +244,145 @@ public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+            // Request the GET_ACCOUNTS permission via a user dialog
+            EasyPermissions.requestPermissions(
+                    this,
+                    "This app needs to access your Google account (via Contacts).",
```
I think this is going to end up in the user interface, right? It should be in `strings.xml` so it can be translated. I also think the message should be a bit more specific. Perhaps "ODK Collect needs access to your Google account to get blank forms and submit filled forms."

shobhitagarwal1612 replied:
Done

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -525,14 +525,18 @@ public void authRequest(Uri url, HashMap<String, String> doneSoFar) {
-            mService = new com.google.api.services.sheets.v4.Sheets.Builder(
+            mSheetsService = new com.google.api.services.sheets.v4.Sheets.Builder(
+                    transport, jsonFactory, credential)
+                    .setApplicationName("ODK-Collect-Sheets")
```
These application names are used in the User Agent header, I believe. I think it makes sense to have a consistent one for all services. Was there a particular reason you used different ones?

shobhitagarwal1612 replied:
I agree with you. I do think it makes more sense to keep only the application name in the User Agent Header. No, there wasn't any particular reason for naming it like this. Just thought it would be better to name differently to detect problem in any particular service during the development. I'll change both of them to "ODK-Collect". Does that sound ok?

lognaturel replied:
Thanks!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+                if (c.getCount() != 1) {
+                    c.close();
+                    try {
+                        throw new FileNotFoundException(drive_fail + Collect.getInstance()
```
I realize you just followed the structure of the previous code but I think `drive_fail` can and should be removed completely from here and from GoogleDriveActivity. It breaks translations and doesn't add any meaningful information.

shobhitagarwal1612 replied:
Done

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -56,104 +59,83 @@
+    private static final String FILE_LIST_KEY = "fileList";
+    private static final String PARENT_ID_KEY = "parentId";
+    private static final String CURRENT_ID_KEY = "currentDir";
+    private static final String drive_fail = "Media Error: ";
```
This and the following constant are never used.

shobhitagarwal1612 replied:
Removed

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+
+        /*  # NOTE #
+         *  Media files are uploaded to Google Drive of user
+         *  All media files are currently saved under folder "odk-uploads"
```
`Open Data Kit/Submissions/formID/`

shobhitagarwal1612 replied:
Done

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+                }
+                c.close();
+
+                // try to get root folder "Open Data Kit"
```
Can you do this folder resolving and potential creation once outside the loop over photos to upload instead of doing it for each item?

You also have exactly the same code for the root folder, the `Submissions` subfolder and the `jrformId` folder. I would recommend creating a series of methods for that called something like `getIDOfFolderWithName(name, inFolder)` and `createAndGetIDOfFolderWithName`.

shobhitagarwal1612 replied:
Sure

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+                com.google.api.services.drive.model.File folder = null;
+
+                for (com.google.api.services.drive.model.File file : files) {
+                    if (folder == null) {
```
Is it possible for `files` to contain a null file? If not, this loop is equivalent to just setting folder equal to the first file, right?

shobhitagarwal1612 replied:
The reason behind doing this looping was to verify that only one instance of the folder exists. If multiple folders are found then it throws an error message eg. "2 folders found". Is this the correct approach?

lognaturel replied:
Interesting, I suppose that is what this does. I would have done something like if the size of `files` is exactly one, set `folder` to its first value. else if it's greater than one, add the error. The loop implies there's some reason to go through individual items.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+                }
+
+                // uploadedPhotos keeps track of the uploaded URL
+                // relative to the path
```
This doesn't look like a relative URL.

shobhitagarwal1612 replied:
Sorry. I will remove this unnecessary comment

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+
+                // uploadedPhotos keeps track of the uploaded URL
+                // relative to the path
+                uploadedPhotos.put(key, "https://drive.google.com/open?id=" + uploadedFileId);
```
URL should be in a constant.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -565,117 +684,83 @@ else if (columnNames.contains(path)) { // if column present in sheet
-                    }
-                }
-            }
+    private String uploadFileToDrive(HashMap<String, String> photosToUpload, String key, com.google.api.services.drive.model.File folder, File toUpload) throws IOException {
```
This parameter list is a bit confusing. There's no reason to pass in the full map of photos to upload since you only need the name of the file. You also don't need a reference to `folder`, only its ID. I think a better set of params would be `File toUpload, String mediaName, String destinationFolderID`

shobhitagarwal1612 replied:
Done. Thanks for noticing that

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -565,117 +684,83 @@ else if (columnNames.contains(path)) { // if column present in sheet
+        //adding meta-data to the file
+        com.google.api.services.drive.model.File fileMetadata =
+                new com.google.api.services.drive.model.File()
+                        .setName(photosToUpload.get(key))
```
I'm not seeing the name when I look at the file details in GDrive. Am I supposed to?

lognaturel replied:
I'm sorry, of course I see the name. I forgot it would be the timestamp name.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -565,117 +684,83 @@ else if (columnNames.contains(path)) { // if column present in sheet
-            // relative to the path
-            uploaded.put(key, picasaPhoto);
-        }
+        FileContent mediaContent = new FileContent("image/jpeg", toUpload);
```
I thought this would fail for pngs but in fact, it looks like PNGs are transferred and given a .jpg extension. That doesn't seem right!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+                    uploadedFileId = uploadFileToDrive(photosToUpload, key, folder, toUpload);
+                } catch (IOException e) {
+                    e.printStackTrace();
+                    mResults.put(id, form_fail + e.getMessage());
```
I think this `form_fail` constant can also be removed. You can do that in a separate pull request if you'd prefer to do a separate cleanup pass for things you left the same as previous code.

shobhitagarwal1612 replied:
I guess I can make this small change in this PR as well

shobhitagarwal1612 replied:
Done

lognaturel replied:
Oops, I should have moved that suggestion up to the first time I saw something like this! Ah well, it's fine in this PR.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java:
```diff
@@ -565,117 +684,83 @@ else if (columnNames.contains(path)) { // if column present in sheet
+            for (com.google.api.services.drive.model.File file : fileList.getFiles()) {
+                files.add(file);
+            }
+            pageToken = fileList.getNextPageToken();
```
At some point we should probably actually use the paging to not block when getting a big list of files. It's ok to leave this for now but would be good to have an issue.

shobhitagarwal1612 replied:
In current scenario, if the user has a large number of files, then it iterates over all the pages and displays all the file in a single list. Are we sure that we want to remove this?
Moreover, it is not affecting the performance if the number of files are less.

lognaturel replied:
Let's not do anything now.

shobhitagarwal1612 replied:
:+1:

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsTask.java:
```diff
@@ -58,7 +59,7 @@ protected String authenticate(Context context, String mGoogleUserName) throws IO
         String scope =
                 "https://picasaweb.google.com/data/";
-        String token = null;
+        String token;
```
This whole authenticate method is now outdated, right?

shobhitagarwal1612 replied:
Yes it is

shobhitagarwal1612 replied:
I think I missed updating the code for auto upload feature!!!

shobhitagarwal1612 replied:
confirmed

lognaturel reviewed (changes requested):
Thanks for taking on this very important project and doing it so well, @shobhitagarwal1612! I've made some comments inline to improve code quality a bit.

I also noticed a few behavior oddities:
- In the Google Drive activity, on rotate, we go from My Drive to Shared With Me or vice-versa.

- Tapping Back in Google Drive goes all the way back to the root.

- When I try to submit to a sheet I don't have access to, the dialog I get has some ugly json. We don't have to do this now but it would be good to think of the top errors users are likely to hit and provide more useful error messages. For example, here's what I see:
<img width="209" alt="screen shot 2017-03-21 at 4 41 40 pm" src="https://cloud.githubusercontent.com/assets/967540/24177406/b16747f8-0e78-11e7-8c2c-a57ac0f354c7.png">
That should be replaced by something like "You don't have access to the submission sheet. Please make sure your survey administrator has given you edit privileges."


- The same image with the same filename can be created multiple times. I attempted to submit to a form I did not have write permissions to, it failed, but the media still got uploaded. Then when my submission succeeded, I ended up with the same image duplicated. <strike>I think perhaps the media should be only be uploaded if the form itself can be submitted? This is a change to existing structure, though, so we can also do that as a separate issue.</strike> The media needs to be uploaded first to get the media url in the sheet. It could be possible to double check the permissions on the sheet first but probably overkill.

yanokwa commented:
This is looking really great, @shobhitagarwal1612! 

I have weird behavior on a Google account where Collect doesn't have OAuth permissions. When I submit the first time, I get some a few dialogs about the form not being uploaded, then the Auth dialog, then there seems to be a delay before the Send Finalized Form is updated. It ends up feeling strange. Can this behavior be improved?

![mar-21-2017 21-20-16](https://cloud.githubusercontent.com/assets/32369/24178020/c129e89a-0e7c-11e7-87a4-38f0884b3cae.gif)

nribeka commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -56,104 +59,83 @@
-
-    private Stack<String> mCurrentPath = new Stack<String>();
-
+    private Stack<String> mCurrentPath = new Stack<>();
```
I think this will fail on version lower than Java 8, right?

nribeka replied:
Is there a minimum Java version in the building how to or readme?

lognaturel replied:
Take a look at the Gradle file -- it uses Java 7 which is the recommendation for Android. Generics type inference was introduced in Java 7 - http://docs.oracle.com/javase/7/docs/technotes/guides/language/type-inference-generic-instance-creation.html

shobhitagarwal1612 committed:
Creates a string resource
Removes ununsed comments
Adds reference link to Drive API v3
Converts member variables to local variables
Removes unused constants

shobhitagarwal1612 committed:
Converts hardcoded string to string resource

shobhitagarwal1612 committed:
Removes duplicate method testNetwork

shobhitagarwal1612 committed:
Replaces e.printStackTrace with Log.e
Adds constant TAG

shobhitagarwal1612 committed:
Modifies the comments

shobhitagarwal1612 committed:
Removes constant drive_error

shobhitagarwal1612 committed:
Removes constant form_error

shobhitagarwal1612 committed:
Creates a constant field

shobhitagarwal1612 committed:
Refactors parameters of uploadFileToDrive

shobhitagarwal1612 commented:
@yanokwa Thanks for noticing such a potential bug. I'll fix it soon

shobhitagarwal1612 committed:
Removes wrong comment

shobhitagarwal1612 committed:
Removes redundant SCOPE

shobhitagarwal1612 committed:
Refactores code to improve code quality

shobhitagarwal1612 committed:
Checks if user has write access before uploading the photos
Modifies Request header for services
Removes multiple calls to API for fetching sheetId and sheetName for multiple form submission

shobhitagarwal1612 commented:
@lognaturel Added a check to confirm if user has write permissions before uploading photos and redundant API calls for  fetching sheetId and sheetName for multiple uploading of forms. 
Also made the necessary code refactoring to improve code quality.

shobhitagarwal1612 committed:
Fixes behaviour of app in case of logging in google account for first time in upload sheets activity

shobhitagarwal1612 committed:
Fixes state loss in case of device rotation in Get blank forms from google drive

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -210,7 +178,7 @@ public void onCreate(Bundle savedInstanceState) {
             MyDrive = false;
 
-            if (testNetwork()) {
+            if (isDeviceOnline()) {
```
😢😢 this should just be `!isDeviceOnline()`. If you want to do a separate pull request to clean up some of these things that you may have seen while going through the code, feel free.

shobhitagarwal1612 replied:
I am pretty sure it should be `isDeviceOnline()`. 
```
/**
     * Checks whether the device currently has a network connection.
     *
     * @return true if the device has a network connection, false otherwise.
     */
    private boolean isDeviceOnline() {
        ConnectivityManager connMgr =
                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
        return (networkInfo != null && networkInfo.isConnected());
    }
```
I have tested on device too and it's working as it should. Can you please recheck if its not working or not on your device a well?

lognaturel replied:
The code works as it is. I was just feeling sad about the empty `if` branch!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java:
```diff
@@ -315,14 +238,149 @@ public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+        JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+        mDriveService = new com.google.api.services.drive.Drive.Builder(
+                transport, jsonFactory, mCredential)
+                .setApplicationName("ODK-Collect-Drive")
```
Remove "Drive"

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -550,18 +558,23 @@ public void authRequest(Uri url, HashMap<String, String> doneSoFar) {
                 startActivityForResult(e.getIntent(), REQUEST_AUTHORIZATION);
             } catch (IOException | GoogleAuthException e) {
+            } catch (MultipleFoldersFoundException e) {
+                e.printStackTrace();
```
Replace with `Log`

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -302,7 +306,7 @@ public void onRequestPermissionsResult(int requestCode,
     @Override
     public void onPermissionsGranted(int requestCode, List<String> list) {
-        // Do nothing.
+//        getResultsFromApi();
```
Were these things you were trying? Better not to leave comments like this unless there's an explanation with them!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -550,18 +558,23 @@ public void authRequest(Uri url, HashMap<String, String> doneSoFar) {
                 //Immediately invalidate so we get a differnet one if we have to try again
                 GoogleAuthUtil.invalidateToken(GoogleSheetsUploaderActivity.this, token);
 
+                getIDOfFolderWithName(GOOGLE_DRIVE_ROOT_FOLDER, null);
```
I believe this could result in some exceptions being thrown and they should be handled. But I think it makes sense to do all the folder getting and creation together in `uploadOneSubmission`. You don't really gain anything by doing it here.

shobhitagarwal1612 committed:
Fixes auto upload feature

lognaturel commented:
It looks like 3a8ae33 breaks form upload for me. I'm now always getting the following error from `GoogleSheetsAbstractUploader.java`:160:
```
400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "Must specify at least one request.",
    "reason" : "badRequest"
  } ],
  "message" : "Must specify at least one request.",
  "status" : "INVALID_ARGUMENT"
}
com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "Must specify at least one request.",
    "reason" : "badRequest"
  } ],
  "message" : "Must specify at least one request.",
  "status" : "INVALID_ARGUMENT"
}
```

shobhitagarwal1612 committed:
modifies the application name for mDriveService in GoogleDriveActivity

shobhitagarwal1612 committed:
Removes unnecessary comments

shobhitagarwal1612 committed:
Fixes 400 bad request while uploading forms

lognaturel commented:
@shobhitagarwal1612 This seems to be working very well now. I think that if you fix that merge conflict, we can merge now. There are a couple small things I'll file as separate issues and it's possible we'll find other small issues as we test for the release but that's totally ok -- it's what code freezes are for!

nribeka commented:
Yay for decreasing the codebase by removing all those "picasa" classes 👍

shobhitagarwal1612 committed:
Fixes spreadsheet name while uploading forms

shobhitagarwal1612 committed:
Merge branch 'master' into get-blank-forms-drive

# Conflicts:
#	collect_app/src/main/res/values/strings.xml

shobhitagarwal1612 commented:
@yanokwa Is the behaviour improved?

lognaturel commented:
👏👏 @shobhitagarwal1612 👏👏 Getting this upgrade/fix in is a big deal!

I'll let @yanokwa take a look at the oauth permission glitch and file an issue if needed.

shobhitagarwal1612 commented:
:smile: It's a pleasure working with all of you 
#ODK
